### PR TITLE
Simple heat consumer implementation

### PR DESCRIPTION
**Pull request by dlohmeier**
This first implementation includes a model that combines the two components heat_exchanger and flow_control. A user can set the desired extracted heat flux and the mass flow through this consumer that can thereby directly connect flow and return system of a district heating network.

### CODE REVIEW

File: src/pandapipes/create.py [hunk 1]
```diff
@@ -1626,6 +1696,78 @@ def create_heat_exchangers(net, from_junctions, to_junctions, diameter_m, qext_w
...
+    :param type: Currently not used - possibility to specify certain types of heat consumers
+    :type type: Iterable(str) or str, default "heat_consumer"
+    :param kwargs: Additional keyword arguments will be added as further columns to the \
+        net["heat_consumer"] table
+    :type kwargs: dict
+    :return: index - The unique IDs of the created heat consumers
+    :rtype: int
+
+    :Example:
+        >>> create_heat_consumers(net,from_junctions=[0, 3], to_junctions=[1, 5], diameter_m=40e-3, qext_w=2000)
```
Comment (EPrade) [inline src/pandapipes/create.py:1741]: Should not two parameter be defined according to line 1751. So qext and maybe mdot?

File: src/pandapipes/create.py [hunk 2]
```diff
@@ -1626,6 +1696,78 @@ def create_heat_exchangers(net, from_junctions, to_junctions, diameter_m, qext_w
...
+    """
+    if np.any(pd.notnull(deltat_k)) or np.any(pd.notnull(treturn_k)):
+        raise NotImplementedError("The models for consumers with fixed temperature difference or "
+                                  "fixed return temperature are not implemented yet.")
+    check_vars = [controlled_mdot_kg_per_s, qext_w, deltat_k, treturn_k]
+    var_sums = np.zeros([4, len(from_junctions)])
+    for i, cv in enumerate(check_vars):
+        var_sums[i] = np.full_like(from_junctions, pd.isnull(cv)).astype(int)
+    if np.any(np.sum(var_sums, axis=0) != 2):
+            raise AttributeError(r"Define exactly two varibales from 'controlled_mdot_kg_per_s', "
```
Comment (EPrade) [inline src/pandapipes/create.py:1751]: typo variables

File: src/pandapipes/create.py (hunk 1 above)
Comment (dlohmeier) [reply src/pandapipes/create.py]: yep, fixed that, thanks.

File: src/pandapipes/create.py (hunk 2 above)
Comment (dlohmeier) [reply src/pandapipes/create.py]: thanks, fixed that.

**Review by EPrade — DISMISSED** [review]

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        :return: No Output.
+        """
+        hs_pit = super().create_pit_branch_entries(net, branch_pit)
+        hs_pit[:, D] = net[cls.table_name()].diameter_m.values
+        hs_pit[:, AREA] = hs_pit[:, D] ** 2 * np.pi / 4
+        hs_pit[:, VINIT] = (net[cls.table_name()].controlled_mdot_kg_per_s.values /
+                            (hs_pit[:, AREA] * hs_pit[:, RHO]))
+        hs_pit[:, ALPHA] = 0
+        hs_pit[:, QEXT] = net[cls.table_name()].qext_w.values
+        hs_pit[:, TEXT] = 293.15
```
Comment (SimonRubenDrauz) [inline src/pandapipes/component_models/heat_consumer_component.py:75]: I think we should move this up to the BranchWZeroLengthComponent

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        # of velocity is allowed during the pipeflow iteration
+        f, t = idx_lookups[cls.table_name()]
+        fc_branch_pit = branch_pit[f:t, :]
+        fc_array = get_component_array(net, cls.table_name())
+        # TODO: this is more precise, but slower:
+        #       np.isin(fc_array[:, cls.MODE], [cls.MF_QE, cls.MF_DT, cls.MF_TR])
+        mdot_controlled = ~np.isnan(fc_array[:, cls.MASS])
+        fc_branch_pit[mdot_controlled, JAC_DERIV_DP] = 0
+        fc_branch_pit[mdot_controlled, JAC_DERIV_DP1] = 0
+        fc_branch_pit[mdot_controlled, JAC_DERIV_DV] = 1
```
Comment (SimonRubenDrauz) [inline src/pandapipes/component_models/heat_consumer_component.py:132]: change to JAC_DERIV_DM

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        Get component input.
+
+        :return:
+        :rtype:
+        """
+        return [("name", dtype(object)),
+                ("from_junction", "u4"),
+                ("to_junction", "u4"),
+                ("controlled_mdot_kg_per_s", "f8"),
+                ("qext_w", 'f8'),
```
Comment (SimonRubenDrauz) [inline src/pandapipes/component_models/heat_consumer_component.py:183]: check order

File: src/pandapipes/test/api/test_components/test_heat_consumer.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        # check for deltat_k and treturn_k given as single values
+        pandapipes.create_heat_consumers(net, juncs[[1, 2]], juncs[[4, 3]], 0.1022,
+                                         deltat_k=30, treturn_k=390)
+
+
+if __name__ == '__main__':
+    pd.set_option("display.width", 1000)
+    pd.set_option("display.max_columns", 45)
+    pd.set_option("display.max_colwidth", 100)
+    pd.set_option("display.max_rows", 200)
```
Comment (SimonRubenDrauz) [inline src/pandapipes/test/api/test_components/test_heat_consumer.py:172]: could you replace this by pytest.main([__file__])

File: src/pandapipes/component_models/heat_consumer_component.py (hunk 4 above)
Comment (dlohmeier) [reply src/pandapipes/component_models/heat_consumer_component.py:132]: We should check which PR to merge first... don't know which order is better :-D

File: src/pandapipes/create.py [hunk 7]
```diff
@@ -1019,6 +1020,75 @@ def create_flow_control(net, from_junction, to_junction, controlled_mdot_kg_per_
     return index
 
 
+def create_heat_consumer(net, from_junction, to_junction, diameter_m, qext_w=None,
+                         controlled_mdot_kg_per_s=None, deltat_k=None, treturn_k=None, name=None,
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply src/pandapipes/create.py:1024]: Hmm... I would rather say that we missed our normal convention in the circulation pump definition... we say qext_w, tfluid_k etc.

[the remark this answers is not in the document]
Comment (dlohmeier) [reply src/pandapipes/create.py:1024]: I could add a format conversion in the circulation pump model and a compatibility assertion, although it's not really nice to change the API, but I think it's better to go with the initial convention here...

File: src/pandapipes/component_models/heat_consumer_component.py (hunk 3 above)
Comment (dlohmeier) [reply src/pandapipes/component_models/heat_consumer_component.py]: Agreed, I moved text and alpha both for heat consumer and heat exchanger.

File: src/pandapipes/component_models/heat_consumer_component.py (hunk 4 above)
Comment (SimonRubenDrauz) [reply src/pandapipes/component_models/heat_consumer_component.py:130]: hehe, yes sure. I guess, it doesn't make a big difference :D

File: src/pandapipes/create.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dlohmeier) [reply src/pandapipes/create.py:1024]: OK, we leave it like that for now, but should consider making a change there.

**Review by SimonRubenDrauz — APPROVED** [review]

**Pull request merged**